include/circt/Dialect/HW/HWAttributes.td

```diff
@@ -11,14 +11,81 @@
 //===----------------------------------------------------------------------===//
 
 /// An attribute to indicate the output file an operation should be emitted to.
-def OutputFileAttr : StructAttr<"OutputFileAttr", HWDialect, [
-  StructFieldAttr<"directory", OptionalAttr<StrAttr>>,
-  StructFieldAttr<"name", OptionalAttr<StrAttr>>,
-  StructFieldAttr<"exclude_from_filelist",
-    DefaultValuedAttr<BoolAttr, "false">>,
-  StructFieldAttr<"exclude_replicated_ops",
-    DefaultValuedAttr<BoolAttr, "true">>,
-]>;
+def OutputFileAttr : AttrDef<HWDialect, "OutputFile"> {
+  let summary = "Ouput file attribute";
+  let description = [{
+    This attribute represents an output file for something which will be
+    printed. The `filename` string is the file to be output to. If `filename`
+    ends in a `/` it is considered an output directory.
+
+    When creating an `OutputFileAttr` the `name` and `directory` should be
+    normalized, with uneeded indirection to the target file removed.  This is
+    so that `OutputFileAttrs` which are pointing at the same file will be
+    equivalent. The `getWithFilename` function will properly handle this
+    process.
```

Comment:
This is pedantic, but I think "normalized" is not a standard definition.  I think what we actually want here is a canonical path which would enable comparisons?  This leads to the question of are we producing canonical paths here and is there a utility inside LLVM to do this already? (More info in a comment on the `normalizeFilename` function.)

Less pedantic, what happens if you have an output file attr pointing at a file and another pointing at a symbolic link to the file? Are these properly merged right now?

Reply:
> This is pedantic, but I think "normalized" is not a standard definition.

Sure I can change it.

> Do you know if what this is doing is going to handle symbolic links?

It's not powerful in the way you describe, it doesn't handle symlinks or even handle an absolute path and relative path to the same file.  It just removes extra dots :)

> would something like llvm::FileCollector::PathCanonicalizer::canonicalize(StringRef srcPath) do the right thing?

The PathCanonicalizer thing looks pretty funky.  I'm going to go on a limb and say this specific utility doesn't look like something we should use. There might be a different utility we can dig up.
 
 > Less pedantic, what happens if you have an output file attr pointing at a file and another pointing at a symbolic link to the file? Are these properly merged right now?

Export verilog will recognize file names based on string equivalence, if they don't match exactly it will reopen the file. Since we are opening these files with `OF_None` I think each time we open the file the old one will be deleted.

Reply:
After an offline conversation, we can implement real canonicalization here using `real_path`, but we decided to go the other way with it.  Instead of canonicalizing paths before creating th `OutputFileAttr` we're going to preserve the user input path. We're going to look at alternative ways of dealing identical output files with different paths in ExportVerilog.